symbolic/src/Data/Macaw/Symbolic/Memory.hs

```diff
@@ -314,30 +371,97 @@ pleatM :: (Monad m, F.Foldable t)
        -> m b
 pleatM s l f = F.foldlM f s l
 
+-- * mapRegionPointers
 
-indexAllocations :: ( MC.MemWidth w
-                    , Ord (WI.SymExpr sym WI.BaseNatType)
-                    )
-                 => IM.IntervalMap (MC.MemWord w) (Allocation sym w)
-                 -> Map.Map (CS.RegValue sym CT.NatType) (Allocation sym w)
-indexAllocations mappedMemory =
-  F.foldl' indexAllocation Map.empty mappedMemory
-  where
-    indexAllocation m alloc =
-      let (base, _) = CL.llvmPointerView (allocationPtr alloc)
-      in Map.insert base alloc m
+-- | Create a function that computes a validity predicate for an LLVMPointer
+-- that may point to the static global memory region.
+--
+-- We represent all of the statically allocated storage in a binary in a single
+-- LLVM array.  This array is sparse, meaning that large ranges of the address
+-- space are not actually mapped.  Whenever we use a pointer into this memory
+-- region, we want to assert that it is inside one of the mapped regions and
+-- that it does not violate any mutability constraints.
+--
+-- The mapped regions are recorded in the MemPtrTable.
+--
+-- We need to be a little careful: if the BlockID of the pointer is definitely
+-- zero, we make a direct assertion.  Otherwise, if the pointer is symbolic, we
+-- have to conditionally assert the range validity.
+--
+-- Note that we pass in an indication of the use of the pointer: if the pointer
+-- is used to write, it must only be within the writable portion of the address
+-- space (which is also recorded in the MemPtrTable).  If the write is
+-- conditional, we must additionally mix in the predicate.
+--
+-- This is intended as a reasonable implementation of the
+-- 'MS.MkGlobalPointerValidityPred'.
+mkGlobalPointerValidityPred :: ( CB.IsSymInterface sym
+                               , MC.MemWidth w
+                               )
+                            => MemPtrTable sym w
+                            -> MS.MkGlobalPointerValidityPred sym w
+mkGlobalPointerValidityPred mpt = \sym puse mcond ptr -> do
+  -- If this is a write, the pointer cannot be in an immutable range (so just
+  -- return False for the predicate on that range).
+  --
+  -- Otherwise, the pointer is allowed to be between the lo/hi range
+  let inMappedRange off (range, mut)
+        | MS.pointerUseTag puse == MS.PointerWrite && mut == CL.Immutable = return (WI.falsePred sym)
+        | otherwise =
+          case range of
+            IM.IntervalCO lo hi -> do
+              lobv <- WI.bvLit sym (memRepr mpt) (fromIntegral lo)
+              hibv <- WI.bvLit sym (memRepr mpt) (fromIntegral hi)
```

Comment:
The `lobv` and `hibv` are identical for each case and could be computed prior to the `case`.

Reply:
Unfortunately, there are no extractors for the `Interval` type to get at the bounds.  I could write one here to factor those out. Do you think it is worth it, given that they need to pattern match anyway?

Reply:
Ahh, that makes sense.  No, I don't think it's worth it in that case.